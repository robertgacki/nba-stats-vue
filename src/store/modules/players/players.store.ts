import { ActionContext, Module } from "vuex";
import {
  State,
  ActionsTypes,
  MutationTypes,
  GettersTypes,
  ApiResponse,
  ApiResponseData,
  Player,
} from "./players.types";
import { RootState } from "@/store";

const state: State = {
  pending: false,
  currentPage: 1,
  totalPages: 0,
  players: [],
};

const mutations = {
  [MutationTypes.SET_PENDING_STATE](state: State, payload: boolean): void {
    state.pending = payload;
  },
  [MutationTypes.SET_CURRENT_PAGE](state: State, payload: number): void {
    if (state.totalPages > payload) {
      state.currentPage = payload;
    } else {
      state.currentPage = state.totalPages;
    }
  },
  [MutationTypes.SET_TOTAL_PAGES](state: State, payload: number): void {
    state.totalPages = payload;
  },
  [MutationTypes.SET_PLAYERS](state: State, payload: ApiResponseData[]): void {
    state.players = payload;
  },
};

const actions = {
  async [ActionsTypes.FETCH](
    { commit }: ActionContext<State, RootState>,
    { page, search }: { page: number; search: string }
  ): Promise<void> {
    commit(MutationTypes.SET_PENDING_STATE, true);

    try {
      const response = await fetch(
        `https://www.balldontlie.io/api/v1/players?page=${page}&search=${search}`
      );
      const { meta, data } = (await response.json()) as ApiResponse;

      commit(MutationTypes.SET_TOTAL_PAGES, meta.total_pages);
      commit(MutationTypes.SET_CURRENT_PAGE, meta.current_page);
      commit(MutationTypes.SET_PLAYERS, data);
    } finally {
      commit(MutationTypes.SET_PENDING_STATE, false);
    }
  },
};

function decoratePlayer(player: ApiResponseData): Player {
  return {
    id: player.id,
    firstName: player.first_name,
    lastName: player.last_name,
    height: player.height_feet
      ? `${player.height_feet}'${player.height_inches}"`
      : "",
    weight: player.weight_pounds ? `${player.weight_pounds} lbs` : "",
    position: player.position,
    teamName: player.team.full_name,
  };
}

const getters = {
  [GettersTypes.GET_PLAYERS](state: State): Player[] {
    return state.players.map(decoratePlayer);
  },
  [GettersTypes.GET_CURRENT_PAGE](state: State): number {
    return state.currentPage;
  },
  [GettersTypes.GET_TOTAL_PAGES](state: State): number {
    return state.totalPages;
  },
  [GettersTypes.GET_PENDING](state: State): boolean {
    return state.pending;
  },
};

const module: Module<State, RootState> = {
  namespaced: true,
  state,
  mutations,
  actions,
  getters,
};

export default module;
