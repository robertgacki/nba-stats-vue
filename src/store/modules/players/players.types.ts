export type State = {
  pending: boolean;
  currentPage: number;
  totalPages: number;
  players: ApiResponseData[];
};

export type Player = {
  id: number;
  firstName: string;
  lastName: string;
  position: string;
  height: string;
  weight: string;
  teamName: string;
};

export type ApiResponse = {
  data: ApiResponseData[];
  meta: ApiResponseMeta;
};

type ApiResponseMeta = {
  total_pages: number;
  current_page: number;
  next_page: number;
  per_page: number;
  total_count: number;
};

export type ApiResponseData = {
  id: number;
  first_name: string;
  last_name: string;
  position: string;
  height_feet: number;
  height_inches: number;
  weight_pounds: number;
  team: ApiResponseDataTeam;
};

type ApiResponseDataTeam = {
  id: number;
  abbreviation: string;
  city: string;
  conference: string;
  division: string;
  full_name: string;
  name: string;
};

export enum MutationTypes {
  SET_PENDING_STATE = "setPendingState",
  SET_CURRENT_PAGE = "setCurrentPage",
  SET_TOTAL_PAGES = "setTotalPages",
  SET_PLAYERS = "setPlayers",
}

export enum ActionsTypes {
  FETCH = "fetch",
}

export enum GettersTypes {
  GET_PLAYERS = "getPlayers",
  GET_TOTAL_PAGES = "getTotalPages",
  GET_CURRENT_PAGE = "getCurrentPage",
  GET_PENDING = "getPending",
}
