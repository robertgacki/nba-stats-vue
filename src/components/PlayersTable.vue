<template>
  <div class="table-responsive">
    <table class="table table-hover my-3">
      <thead>
        <tr>
          <th scope="col">Player name</th>
          <th scope="col" class="text-center">Position</th>
          <th scope="col" class="text-center">Height</th>
          <th scope="col" class="text-center">Weight</th>
          <th scope="col">Team name</th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="player in players" :key="player.id">
          <td>
            <router-link
              :to="{ name: 'Player', params: { id: player.id } }"
              class="link-primary"
              >{{ player.firstName }} {{ player.lastName }}</router-link
            >
          </td>
          <td class="text-center">{{ player.position }}</td>
          <td class="text-center">{{ player.height }}</td>
          <td class="text-center">{{ player.weight }}</td>
          <td>{{ player.teamName }}</td>
        </tr>
      </tbody>
    </table>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";

import { Player } from "@/store/modules/players/players.types";

export default defineComponent({
  props: {
    players: Array as PropType<Player[]>,
  },
});
</script>
