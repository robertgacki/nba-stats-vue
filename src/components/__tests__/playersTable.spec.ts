import PlayersTable from "@/components/PlayersTable.vue";
import { shallowMount, RouterLinkStub } from "@vue/test-utils";
import { Player } from "@/store/modules/players/players.types";

describe("PlayersTable", () => {
  describe("snapshots", () => {
    it("should render proper snapshot when players prop is empty", () => {
      const wrapper = shallowMount(PlayersTable, {
        global: {
          stubs: {
            RouterLink: RouterLinkStub,
          },
        },
      });
      wrapper.setProps({ players: [] });

      expect(wrapper.vm.$el).toMatchSnapshot();
    });

    it("should render proper snapshot when players prop is not empty", () => {
      const wrapper = shallowMount(PlayersTable, {
        global: {
          stubs: {
            RouterLink: RouterLinkStub,
          },
        },
        props: {
          players: [
            {
              id: 2,
              firstName: "Michael",
              lastName: "Jordan",
              height: `7'3"`,
              weight: "200 lbs",
              position: "F",
              teamName: "Chicago Bulls",
            },
            {
              id: 3,
              firstName: "Marcin",
              lastName: "Gortat",
              height: `8'5"`,
              weight: "235 lbs",
              position: "C",
              teamName: "New York Knicks",
            },
          ] as Player[],
        },
      });

      expect(wrapper.vm.$el).toMatchSnapshot();
    });
  });
});
