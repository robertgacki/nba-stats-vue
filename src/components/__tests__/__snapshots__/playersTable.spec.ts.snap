// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayersTable snapshots should render proper snapshot when players prop is empty 1`] = `
<div
  class="table-responsive"
>
  <table
    class="table table-hover my-3"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Player name
        </th>
        <th
          class="text-center"
          scope="col"
        >
          Position
        </th>
        <th
          class="text-center"
          scope="col"
        >
          Height
        </th>
        <th
          class="text-center"
          scope="col"
        >
          Weight
        </th>
        <th
          scope="col"
        >
          Team name
        </th>
      </tr>
    </thead>
    <tbody>
      
      
    </tbody>
  </table>
</div>
`;

exports[`PlayersTable snapshots should render proper snapshot when players prop is not empty 1`] = `
<div
  class="table-responsive"
>
  <table
    class="table table-hover my-3"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Player name
        </th>
        <th
          class="text-center"
          scope="col"
        >
          Position
        </th>
        <th
          class="text-center"
          scope="col"
        >
          Height
        </th>
        <th
          class="text-center"
          scope="col"
        >
          Weight
        </th>
        <th
          scope="col"
        >
          Team name
        </th>
      </tr>
    </thead>
    <tbody>
      
      <tr>
        <td>
          <a
            class="link-primary"
          >
            Michael Jordan
          </a>
        </td>
        <td
          class="text-center"
        >
          F
        </td>
        <td
          class="text-center"
        >
          7'3"
        </td>
        <td
          class="text-center"
        >
          200 lbs
        </td>
        <td>
          Chicago Bulls
        </td>
      </tr>
      <tr>
        <td>
          <a
            class="link-primary"
          >
            Marcin Gortat
          </a>
        </td>
        <td
          class="text-center"
        >
          C
        </td>
        <td
          class="text-center"
        >
          8'5"
        </td>
        <td
          class="text-center"
        >
          235 lbs
        </td>
        <td>
          New York Knicks
        </td>
      </tr>
      
    </tbody>
  </table>
</div>
`;
